enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,244 @@
+
+With this in mind, this enhancement proposes adding the `failForwardUpgrades` toggle to an already watched namespaced resource, the OperatorGroup.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  failForwardUpgrades: true
+```
+
+With the toggle enabled, OLM will allow operators to "fail forward" if such upgrades are supported by the operator. If the toggle is disabled, OLM exhibits existing behavior.
+
+#### Supporting Fail Forward Upgrades
+
+This enhancement proposes the introduction of the `fail-forward-supported` [Bundle Property](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#properties). When the resolver determines the set of operators to install, OLM can place constraints on a bundle that must be met in the next generation of installables.
+
+To allow installs past the failed install, a catalog curator/operator author would need to rebuild the operator's upgrade graph to include a CSV that:
+- Replaces the version of the operator that failed to install.
+- Includes the `fail-forward-supported` Bundle property.
+
+A bundle with the `fail-forward-supported` property can be seen below:
+
+```yaml=
+image: quay.io/foo/bar:baz
+name: foo-operator.v1.0.1
+package: bar
+properties:
+- type: olm.gvk
+  value:
+    group: foo.io
+    kind: bar
+    version: v1alpha1
+- type: olm.package
+  value:
+    packageName: stable
+    version: 1.0.1
+- type: olm.failForward 
+  value:
+    supported: true # This bundle supports "failing forward".
+schema: olm.bundle
+```
+
+With this property present, OLM can then communicate to the resolver when a particular operator failed to install. The resolver can then check if an upgrade exists for the operator that supports "failing forward". Let's consider each of the scenarios OLM should Support:
+
+##### InstallPlan Failure
+
+If the InstallPlan has failed, the set of operators sent to the resolver should include:
+-- Each CSV that successfully installed
+-- Each CSV that was not installed (OLM will include the "fail forward" constraint).
+
+>Note: It might make sense just to recreate the InstallPlan here if the operatorGroup is configured to "fail forward", please see [Open Questions](#open-questions).
+
+##### CSV Failure
+If a CSV in the namespace is in the FAILED state, the set of operators sent to the resolver should include:
+-- Each CSV that successfully installed.
+-- Each CSV that failed to install (OLM wil include the "fail forward" constraint)
+-- Each CSV in the "Pending State" (OLM wil include the "fail forward" constraint)
+
+>Note: How does OLM determine that a CSV is "stuck" in the pending state? Perhaps opting into "fail forward" upgrades should include configurable timeout, please see [Open Questions](#open-questions).
```

Comment:
If we treating CSVs that are "stuck" in pending as failed from the perspective of failing forward, maybe it makes sense to simply put them in the Failed state? Probably not worth a backwards-incompatible behavioral change but it would make CSV state management more clear. 

A lot of times a CSV is stuck in Pending due to a lack of permissions. In theory, someone could come along and grant additional permissions, but the install has effectively failed.

Reply:
Technically not a backward-incompatible behavior change since the user needs to opt-into fail forward upgrades but I agree with the sentiment. Moving it to the failed phase first seems reasonable.